### PR TITLE
Separate pending users from blocked users

### PR DESCRIPTION
**Pull request by peregrineshahin**

### CODE REVIEW

File: server/fishtest/templates/user.mak [hunk 1]
```diff
@@ -83,21 +83,38 @@
...
-        <input
-          type="checkbox"
-          class="form-check-input"
-          id="blocked"
-          name="blocked"
-          value="True"
-          ${checked}
-        />
-      </div>
+      if user['pending']:
```
Comment (ppigazzini) [inline server/fishtest/templates/user.mak:86]:
```python
if "pending" in user and user["pending"]:
```
or if "pending" is enforced to be a bool:
```python
if user.get("pending") is not None:
```

**Comment by ppigazzini** [comment]
> Maybe this also needs to add `pending` to all old database users?

I think so, the script is already in place and it needs a simple update.

File: server/fishtest/templates/user.mak [hunk 2]
```diff
@@ -83,21 +83,34 @@
...
-        <input
-          type="checkbox"
-          class="form-check-input"
-          id="blocked"
-          name="blocked"
-          value="True"
-          ${checked}
-        />
-      </div>
+      if 'pending' in user and user['pending']:
```
Comment (ppigazzini) [inline server/fishtest/templates/user.mak:86]: Wrong `if-elif-else` mako syntax and uneven formatting:
```diff
diff --git a/server/fishtest/templates/user.mak b/server/fishtest/templates/user.mak
index 6c31e75..3b15619 100644
--- a/server/fishtest/templates/user.mak
+++ b/server/fishtest/templates/user.mak
@@ -82,35 +82,34 @@
           <i class="fa-solid fa-lg fa-eye pe-none" style="width: 30px"></i>
         </span>
       </div>
+    % elif 'pending' in user and user['pending']:
+      <div class="mb-3 form-check">
+        <label class="form-check-label" for="pending">Pending</label>
+        <input
+          type="checkbox"
+          class="form-check-input"
+          id="pending"
+          name="pending"
+          value="True"
+          checked
+        />
+      </div>
     % else:
-      if 'pending' in user and user['pending']:
-        <div class="mb-3 form-check">
-          <label class="form-check-label" for="pending">Pending</label>
-          <input
-            type="checkbox"
-            class="form-check-input"
-            id="pending"
-            name="pending"
-            value="True"
-            checked
-          />
-        </div>
-      else:
-        <%
-          blocked = user['blocked'] if 'blocked' in user else False
-          blocked_checked = 'checked' if blocked else ''
-        %>
-        <div class="mb-3 form-check">
-          <label class="form-check-label" for="blocked">Blocked</label>
-          <input
-            type="checkbox"
-            class="form-check-input"
-            id="blocked"
-            name="blocked"
-            value="True"
-            ${blocked_checked}
-          />
-        </div>
+      <%
+        blocked = user['blocked'] if 'blocked' in user else False
+        blocked_checked = 'checked' if blocked else ''
+      %>
+      <div class="mb-3 form-check">
+        <label class="form-check-label" for="blocked">Blocked</label>
+        <input
+          type="checkbox"
+          class="form-check-input"
+          id="blocked"
+          name="blocked"
+          value="True"
+          ${blocked_checked}
+        />
+      </div>
     % endif
```

**Comment by peregrineshahin** [comment]
How error?, didn't we agree that we need to add pending to all users?
> Error after setting the "Blocked" checkbox and pushing "Submit":
> 
> ```
> Sep 26 17:38:15 dfts-0 pserve[3078]:   File "/home/usr00/fishtest/server/fishtest/views.py", line 629, in user
> Sep 26 17:38:15 dfts-0 pserve[3078]:     change_pending = user_data["pending"] != "pending" in request.POST
> Sep 26 17:38:15 dfts-0 pserve[3078]:                      ~~~~~~~~~^^^^^^^^^^^
> Sep 26 17:38:15 dfts-0 pserve[3078]: KeyError: 'pending'
> ```

File: server/fishtest/actiondb.py [hunk 3]
```diff
@@ -78,6 +78,12 @@
         "user": str,
         "message": union("blocked", "unblocked"),
     },
+    {
+        "action": "accept_user",
+        "username": str,
+        "user": str,
+        "message": union("accept", "accept"),
```
[the remark this answers is not in the document]
Comment (peregrineshahin) [reply server/fishtest/actiondb.py:85]: Thanks I meant "pending".
> `union("accept", "accept")` is the same as just `"accept"`. Perhaps you mean something like `union("accepted", "not accepted")`?

**Comment by ppigazzini** [comment]
user DB collection updated with the field "pending", DEV updated with the PR, registered 2 new users, there are both "pending" and "blocked"

image

File: server/fishtest/templates/contributors.mak [hunk 4]
```diff
@@ -5,11 +5,11 @@
 %>
 
 <script>
-  document.title = 'Users${" - Top Month" if "monthly" in request.url else ""} | Stockfish Testing';
+  document.title = 'contributors${" - Top Month" if "monthly" in request.url else ""} | Stockfish Testing';
```
Comment (ppigazzini) [inline server/fishtest/templates/contributors.mak:8]: All titles start with upper letter, so "Contributors"

**Comment by peregrineshahin** [comment]
should be fixed

**Comment by ppigazzini** [comment]
DEV updated

**Comment by ppigazzini** [comment]
unblock and unpending seem broken.

**Comment by peregrineshahin** [comment]
I pushed a fix, maybe this hacks it. basically, if checkboxes are unchecked, HTML forums don't include them in the post request as opposed to sending them with a "False" value.

**Comment by ppigazzini** [comment]
Unblock works

Unpending fails with:
```
Dec 08 13:49:00 dfts-0 pserve[6648]: 2023-12-08 13:49:00,218 ERROR [waitress][waitress-2] Exception while serving /user/newuser001
Dec 08 13:49:00 dfts-0 pserve[6648]: Traceback (most recent call last):
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/tweens.py", line 13, in _error_handler
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = request.invoke_exception_view(exc_info)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/view.py", line 786, in invoke_exception_view
Dec 08 13:49:00 dfts-0 pserve[6648]:     raise HTTPNotFound
Dec 08 13:49:00 dfts-0 pserve[6648]: pyramid.httpexceptions.HTTPNotFound: The resource could not be found.
Dec 08 13:49:00 dfts-0 pserve[6648]: During handling of the above exception, another exception occurred:
Dec 08 13:49:00 dfts-0 pserve[6648]: Traceback (most recent call last):
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/waitress/channel.py", line 428, in service
Dec 08 13:49:00 dfts-0 pserve[6648]:     task.service()
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/waitress/task.py", line 168, in service
Dec 08 13:49:00 dfts-0 pserve[6648]:     self.execute()
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/waitress/task.py", line 434, in execute
Dec 08 13:49:00 dfts-0 pserve[6648]:     app_iter = self.channel.server.application(environ, start_response)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/waitress/proxy_headers.py", line 64, in translate_proxy_headers
Dec 08 13:49:00 dfts-0 pserve[6648]:     return app(environ, start_response)
Dec 08 13:49:00 dfts-0 pserve[6648]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/router.py", line 270, in __call__
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = self.execution_policy(environ, self)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/router.py", line 276, in default_execution_policy
Dec 08 13:49:00 dfts-0 pserve[6648]:     return router.invoke_request(request)
Dec 08 13:49:00 dfts-0 pserve[6648]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/router.py", line 245, in invoke_request
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = handle_request(request)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/tweens.py", line 43, in excview_tween
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = _error_handler(request, exc)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/tweens.py", line 17, in _error_handler
Dec 08 13:49:00 dfts-0 pserve[6648]:     reraise(*exc_info)
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/util.py", line 733, in reraise
Dec 08 13:49:00 dfts-0 pserve[6648]:     raise value
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/tweens.py", line 41, in excview_tween
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = handler(request)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/router.py", line 143, in handle_request
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = _call_view(
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/view.py", line 674, in _call_view
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = view_callable(context, request)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/viewderivers.py", line 427, in rendered_view
Dec 08 13:49:00 dfts-0 pserve[6648]:     result = view(context, request)
Dec 08 13:49:00 dfts-0 pserve[6648]:              ^^^^^^^^^^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/env/lib/python3.11/site-packages/pyramid/viewderivers.py", line 141, in _requestonly_view
Dec 08 13:49:00 dfts-0 pserve[6648]:     response = view(request)
Dec 08 13:49:00 dfts-0 pserve[6648]:                ^^^^^^^^^^^^^
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/fishtest/views.py", line 630, in user
Dec 08 13:49:00 dfts-0 pserve[6648]:     request.actiondb.accept_user(
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/fishtest/actiondb.py", line 288, in accept_user
Dec 08 13:49:00 dfts-0 pserve[6648]:     self.insert_action(
Dec 08 13:49:00 dfts-0 pserve[6648]:   File "/home/usr00/fishtest/server/fishtest/actiondb.py", line 303, in insert_action
Dec 08 13:49:00 dfts-0 pserve[6648]:     raise Exception("Validation failed with error '{}'".format(ret))
Dec 08 13:49:00 dfts-0 pserve[6648]: Exception: Validation failed with error 'action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and
```
l

**Comment by vdbergh** [comment]
I'll have a look.

**Comment by vdbergh** [comment]
This is the full error message.
```
Exception: Validation failed with error 'action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['user'] is not in the schema and action['message'] (value:'accepted') is not equal to 'blocked' and action['message'] (value:'accepted') is not equal to 'unblocked' and action['user'] is not in the schema and action['message'] (value:'accepted') is not equal to 'accept' and action['message'] (value:'accepted') is not equal to 'pending''
```
Staring at it now.

**Comment by ppigazzini** [comment]
It works this alternative as well:

```diff
diff --git a/server/fishtest/actiondb.py b/server/fishtest/actiondb.py
index 2f89c74..0328be8 100644
--- a/server/fishtest/actiondb.py
+++ b/server/fishtest/actiondb.py
@@ -89,7 +89,7 @@ schema = union(
         "action": "accept_user",
         "username": str,
         "user": str,
-        "message": union("accept", "pending"),
+        "message": union("accepted", "pending"),
     },
 )
```

**Comment by ppigazzini** [comment]
- open a pending user in two tabs
- uncheck "pending" in the first tab and press "Submit"
- press "Submit" in the second tab, the user now is returned in "pending" state

If I'm not wrong the "pending" -> "accepted" should works only one way.

**Comment by ppigazzini** [comment]
Loading a clean DB and trying some one/way code, adding the "pending" key on the user creation and deleting the key when Accepting

**Comment by peregrineshahin** [comment]
> Loading a clean DB and trying some one/way code, adding the "pending" key on the user creation and deleting the key when Accepting

I added the code to make it one way, I don't think deleting key from the database is a good idea, it also invalidates the schema

**Comment by ppigazzini** [comment]
> > Loading a clean DB and trying some one/way code, adding the "pending" key on the user creation and deleting the key when Accepting
> 
> I added the code to make it one way, I don't think deleting key from the database is a good idea, it also invalidates the schema

It was a good idea, the DEV scripts need an update to work with this PR.

```
2023-12-08T16:17:04.215+0100    index: &idx.IndexDocument{Options:primitive.M{"name":"worker_name_1", "unique":true, "v":2}, Key:primitive.D{primitive.E{Key:"worker_name", Value:1}}, PartialFilterExpression:primitive.D(nil)}
2023-12-08T16:19:01.671+0100    1296748 document(s) restored successfully. 0 document(s) failed to restore.
Unable to obtain the port number. Error: -1
user['pending'] is missing
Traceback (most recent call last):
  File "<stdin>", line 16, in <module>
  File "/home/usr00/fishtest/server/fishtest/userdb.py", line 144, in save_user
    validate_user(user)
  File "/home/usr00/fishtest/server/fishtest/userdb.py", line 30, in validate_user
    assert False
AssertionError
Unable to obtain the port number. Error: -1
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/home/usr00/fishtest/server/fishtest/userdb.py", line 112, in add_user_group
    user = self.find(username)
```

**Comment by ppigazzini** [comment]
Looks good on DEV.
I need to check if the nginx conf needs an update.

Steps:
- [ ] merge the PR
- [ ] update PROD
- [ ] update the "users" collection adding the "pending" key (before the merge is forbidden by the old schema)
- [ ] update the DEV script
- [ ] update the server setup script

**Comment by vdbergh** [comment]
IIUC one cannot go back fror `accepted` to `pending`. If that is so then in the drop down menu in the event log I think one should just have `Accept User`. Not `Pend/Accept User`.

**Comment by vdbergh** [comment]
```
            elif "pending" in request.POST and user_data["pending"]:
                user_data["pending"] = False
                request.userdb.last_pending_time = 0
                request.actiondb.accept_user(
                    username=userid,
                    user=user_name,
                    message="pending" if user_data["pending"] else "accepted",
                )
```
It seems to me that the message can never be `pending`.... So currently the message is redundant in the schema for the actiondb.

This being said: I wonder however if one can create an `accept_user` event on user registration. This could then be displayed with message `pending`.

**Comment by ppigazzini** [comment]
> IIUC one cannot go back from `accepted` to `pending`. If that is so then in the drop down menu in the event log I think one should just have `Accept User`. Not `Pend/Accept User`.

Fixed.
@peregrineshahin get the update version from github (eg `git fetch -p upstream; git reset --hard upstream/blockedUsers`)

```diff
diff --git a/server/fishtest/templates/actions.mak b/server/fishtest/templates/actions.mak
index 1dae4166..e3eb7bd0 100644
--- a/server/fishtest/templates/actions.mak
+++ b/server/fishtest/templates/actions.mak
@@ -24,7 +24,7 @@
       <option value="finished_run">Finished Runs</option>
       <option value="block_user">Block/Unblock User</option>
       <option value="block_worker">Block/Unblock Worker</option>
-      <option value="accept_user">Pend/Accept User</option>
+      <option value="accept_user">Accept User</option>
       <option value="upload_nn">Upload NN file</option>
       <option value="failed_task">Failed Tasks</option>
       <option value="crash_or_time">Crashes or Time losses</option>
```

**Comment by peregrineshahin** [comment]
One moment I want to make another cleanup

**Comment by peregrineshahin** [comment]
Thanks!

> ```python
>             elif "pending" in request.POST and user_data["pending"]:
>                 user_data["pending"] = False
>                 request.userdb.last_pending_time = 0
>                 request.actiondb.accept_user(
>                     username=userid,
>                     user=user_name,
>                     message="pending" if user_data["pending"] else "accepted",
>                 )
> ```
> 
> It seems to me that the message can never be `pending`.... So currently the message is redundant in the schema for the actiondb.
> 
> This being said: I wonder however if one can create an `accept_user` event on user registration. This could then be displayed with message `pending`.

**Comment by peregrineshahin** [comment]
wait there is something wrong about the logic.. I think the unpending is broken no?

**Comment by peregrineshahin** [comment]
should be fixed now

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by peregrineshahin** [comment]
it works now.

**Comment by ppigazzini** [comment]
DEV updated.

**Review by ppigazzini — APPROVED** [review]
Looks good to me.

Steps:
- [x] merge the PR
- [x] update PROD
- [x] update the "users" collection adding the "pending" key (before the merge is forbidden by the old schema)
- [x] update the nginx config
- [x] update the DEV util scripts
- [ ] update the server setup scripts

**Comment by ppigazzini** [comment]
Great, delta_update_users.py crashed on PROD while adding the "pending" key. Start debugging.

```
 ${VENV}/bin/python3 delta_update_users.py
Unable to obtain the port number. Error: -1
update scan
load deltas:
len(deltas)=148490
next(iter(deltas))='6573b921f09ce1261f12322a'
Successfully updated 2286 users
Successfully updated 120 top month users
user['email'] (value:'') is not of type 'email': The email address is not valid. It must have exactly one @-sign.
Traceback (most recent call last):
  File "/home/fishtest/fishtest/server/utils/delta_update_users.py", line 289, in <module>
    main()
  File "/home/fishtest/fishtest/server/utils/delta_update_users.py", line 283, in main
    cleanup_users(rundb)
  File "/home/fishtest/fishtest/server/utils/delta_update_users.py", line 207, in cleanup_users
    rundb.userdb.save_user(u)
  File "/home/fishtest/fishtest/server/fishtest/userdb.py", line 144, in save_user
    validate_user(user)
  File "/home/fishtest/fishtest/server/fishtest/userdb.py", line 30, in validate_user
    assert False
AssertionError
```

**Comment by peregrineshahin** [comment]
It says it's related to email not being there. Maybe some users in the top contributers do not have an email.

**Comment by vdbergh** [comment]
This is the culprit.
```
user['email'] (value:'') is not of type 'email': The email address is not valid. It must have exactly one @-sign
```

**Comment by vdbergh** [comment]
If email can be '' then one can put `union(email, '')` in the schema. Then it will validate the email if present but it will also allow an empty email.

**Comment by ppigazzini** [comment]
> This is the culprit.
> 
> ```
> user['email'] (value:'') is not of type 'email': The email address is not valid. It must have exactly one @-sign
> ```

I do know. The problem is that old users could have a empty string or a random string as email key.
On DEV I didn't view the problem because I rewrote the email as user001@example.org

**Comment by peregrineshahin** [comment]
Could it be easier to fill them up manually?

**Comment by vdbergh** [comment]
@peregrineshahin I guess one does not know the email of some old users.

**Comment by ppigazzini** [comment]
Well, with the updated schema `union("", email)` I'm getting the wrong user/email when the script crashes :)

**Comment by peregrineshahin** [comment]
Yeah probably a random string + @example.com
I hope that we don't fill with the same string, because I have a PR open that prevents registration with the same email.

**Comment by ppigazzini** [comment]
Our userbase has a great fantasy in hacking the email value :)

**Comment by vdbergh** [comment]
> Well, with the updated schema `union("", email)` I'm getting the wrong email value when the script crashes :)

I don't understand. The message should essentially be
```
bad_email is not equal to '' and bad_email is not of type 'email'
```

**Comment by peregrineshahin** [comment]
> Our userbase has a great fantasy in hacking the email value :)

Many users did not read the message that approval is manual, and they insist on registeration with the same email, I would remove their duplicate accounts and keep the last registered one.

**Comment by ppigazzini** [comment]
> > Well, with the updated schema `union("", email)` I'm getting the wrong email value when the script crashes :)
> 
> I don't understand. The error message should essentially be
> 
> ```
> bad_email is not equal to '' and bad_email is not of type 'email'
> ```

I'm dropping the "numeric string"
I'm replacing the dot with @ when finding "username.domainname"
I'm adding example.org when finding "username@"

**Comment by vdbergh** [comment]
I see. I guess the second one may be problematic is the domain name misses a TLD.

**Comment by ppigazzini** [comment]
Done, only 5 emails to be fixed.

**Comment by ppigazzini** [comment]
We have 30 users with an empty email field.

**Comment by vdbergh** [comment]
That's not too bad.

**Comment by ppigazzini** [comment]
Completed the post merge tasks, thank you @peregrineshahin :)

**Pull request merged**